src/hb-ot-shape.cc

```diff
@@ -923,15 +923,19 @@ zero_mark_width (hb_glyph_position_t *pos)
 static inline void
 zero_mark_widths_by_gdef (hb_buffer_t *buffer, bool adjust_offsets)
 {
+  unsigned int prev_cluster = (unsigned int) -1;
   unsigned int count = buffer->len;
   hb_glyph_info_t *info = buffer->info;
   for (unsigned int i = 0; i < count; i++)
+  {
     if (_hb_glyph_info_is_mark (&info[i]))
     {
-      if (adjust_offsets)
+      if (adjust_offsets && likely (info[i].cluster == prev_cluster))
```

Comment:
This can't be good. Fails with other cluster levels. I need to study why you are doing this. Typically we just insert a dotted-circle if there's no base before a mark. But that can only happen at start of text.

Reply:
The motivation for this is that the Apple Color Emoji case here (the couple-with-skintones-holding-hands sequence) relies on substituting the first "person" glyph (MAN or WOMAN) with a version that has GDEF class = 3, and expects that to result in zeroing its advance. But it must not be repositioned leftwards when its advance gets zeroed, because it isn't supposed to overprint the preceding character; it's supposed to stay in place so that the second "person" glyph overprints it.

But you're right, this would be a problem with other cluster levels...

(And there's no dotted circle because although this is now at the start of text in the testcase, it wasn't a mark glyph at the time when dotted-circle insertion would have happened. It only becomes a mark during morx substitution.)

It would've been easier, perhaps, if they'd simply given these holding-hands people glyphs half the advance width each, instead of creating glyphs with the full width but the person image only in half of the space, and then overprinting the two glyphs at the same position -- using the GDEF class to trigger the zeroing of the first glyph's advance.

Reply:
Right...

I feel like we should disable that adjustment for morx/kerx somehow. That is, let's adjust the logic here:
```
  plan.zero_marks = script_zero_marks &&
                    !plan.apply_kerx &&
                    (!plan.apply_kern
#ifndef HB_NO_OT_KERN
                     || !hb_ot_layout_has_machine_kerning (face)
#endif
                    );
  plan.has_gpos_mark = !!plan.map.get_1_mask (HB_TAG ('m','a','r','k'));

  plan.adjust_mark_positioning_when_zeroing = !plan.apply_gpos &&
                                              !plan.apply_kerx &&
                                              (!plan.apply_kern
#ifndef HB_NO_OT_KERN
                                               || !hb_ot_layout_has_cross_kerning (face)
#endif
                                              );

  plan.fallback_mark_positioning = plan.adjust_mark_positioning_when_zeroing &&
                                   script_fallback_mark_positioning;
```

Reply:
OK, I think we can just disable adjust_mark_positioning_when_zeroing (but not fallback_mark_positioning) if we're using apply_morx. This resolves the problem here, and so far doesn't seem to break expectations anywhere else. I've updated the PR to use this approach.

Also added a testcase for the sequence involved here. Maybe @ebraminio can check if older versions of the font supported it (might be quite a recent addition?), and if so add the appropriate entries for wider coverage.

Reply:
I don't have access to older version of the macOS fonts, I used to access those in the CI bots but Circle CI won't let me log into it anymore as they have added some country based blocks.

Reply:
Thanks very much for the fix!

> Also added a testcase for the sequence involved here. Maybe @ebraminio can check if older versions of the font supported it (might be quite a recent addition?), and if so add the appropriate entries for wider coverage.

According to https://blog.emojipedia.org/ios-13-2-emoji-changelog/ the first release that had the mixed holding hands emoji was iOS 13 (September 2019). Mac OS 10.15 came out after (October 2019) that so maybe 10.15 Apple Color Emoji may have it. I only have archived copies of 10.10 to 10.13 here and in 10.13 definitely the sequence does not exist as a combination.